clab/clab.go

```diff
@@ -517,32 +531,126 @@ func (c *CLab) scheduleNodes(ctx context.Context, maxWorkers int,
 					continue
 				}
+
+				// enter the create stage
+				err = dm.EnterStage(node.Config().ShortName, types.WaitForCreate)
+				if err != nil {
+					log.Error(err)
+				}
+
 				// Deploy
 				err = node.Deploy(ctx, &nodes.DeployParams{})
 				if err != nil {
-					log.Errorf("failed deploy phase for node %q: %v", node.Config().ShortName, err)
+					log.Errorf("failed deploy stage for node %q: %v", node.Config().ShortName, err)
 					continue
 				}
 
+				// we need to update the node's state with runtime info (e.g. the mgmt net ip addresses)
+				// before continuing with the post-deploy stage (for e.g. certificate creation)
+				err = node.UpdateConfigWithRuntimeInfo(ctx)
+				if err != nil {
+					log.Errorf("failed to update node runtime information for node %s: %v", node.Config().ShortName, err)
+				}
+
+				dm.SignalDone(node.GetShortName(), types.WaitForCreate)
+
+				err = dm.EnterStage(node.GetShortName(), types.WaitForCreateLinks)
+				if err != nil {
+					log.Error(err)
+				}
+
 				err = node.DeployLinks(ctx)
 				if err != nil {
 					log.Errorf("failed deploy links for node %q: %v", node.Config().ShortName, err)
 					continue
 				}
 
-				// signal to dependency manager that this node is done with creation
-				dm.SignalDone(node.Config().ShortName, dependency_manager.NodeStateCreated)
+				// DeployLinks does not necessarily create all the links already.
+				// veth links might be created by the peer node, if it was not already
+				// deployed properly. Hence we need to wait for all the links to be created.
+				// we just wait if there is actually a dependency on this state, otherwise
+				// we head on.
+				count, err := dm.GetDependerCount(node.GetShortName(), types.WaitForCreateLinks)
+				if err != nil {
+					log.Error(err)
+				}
+				if count > 0 {
+					node.WaitForAllLinksCreated()
+				}
+				dm.SignalDone(node.GetShortName(), types.WaitForCreateLinks)
+
+				// start config stage
+				err = dm.EnterStage(node.GetShortName(), types.WaitForConfigure)
+				if err != nil {
+					log.Error(err)
+				}
+
+				// if postdeploy should be skipped we do not call it
+				if !skipPostDeploy {
+					err = node.PostDeploy(ctx, &nodes.PostDeployParams{Nodes: c.Nodes})
+					if err != nil {
+						log.Errorf("failed to run postdeploy task for node %s: %v", node.Config().ShortName, err)
+					}
+				}
+
+				dm.SignalDone(node.GetShortName(), types.WaitForConfigure)
+
+				// run execs
+				err = node.RunExecFromConfig(ctx, execCollection)
+				if err != nil {
+					log.Errorf("failed to run exec commands for %s: %v", node.GetShortName(), err)
+				}
+
+				// health state processing
+				count, err = dm.GetDependerCount(node.GetShortName(), types.WaitForHealthy)
+				if err != nil {
+					log.Error(err)
+				}
+				if count > 0 {
+					// if there is a dependecy on the healthy state of this node, enter the checking procedure
+					for {
+						healthy, err := node.GetRuntime().IsHealthy(ctx, node.Config().LongName)
```

Comment:
@steiler it feels a bit off that IsHealthy is a method of a runtime and not the node itself.
Doesn't it make more sense to have `node.IsHealthy()`?

Reply:
Agree. We even have a node.GetState() but it does not resembel whats required here. I'll take a look if we can unify it or if we need a seperate function.

Reply:
alright did introduce that